include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh

```diff
@@ -1178,107 +1174,120 @@ class open_addressing_ref_impl {
-                  break;
-                }
-                default: {
-                  break;
+          this->probing_scheme_(probing_tile, probe_key, this->storage_ref_.bucket_extent());
+
+        bool running                      = true;
+        [[maybe_unused]] bool found_match = false;
+
+        bool equals[buffer_size];
+        uint32_t exists[buffer_size];
+
+        while (active_flushing_tile.any(running)) {
+          if (running) {
+            // TODO atomic_ref::load if insert operator is present
+            auto const bucket_slots = this->storage_ref_[*probing_iter];
+
+#pragma unroll buffer_size
+            for (int32_t i = 0; i < bucket_size; ++i) {
+              equals[i] = false;
+              if (running) {
+                // inspect slot content
+                switch (this->predicate_.operator()<is_insert::NO>(
+                  probe_key, this->extract_key(bucket_slots[i]))) {
+                  case detail::equal_result::EMPTY: {
+                    running = false;
+                    break;
+                  }
+                  case detail::equal_result::EQUAL: {
+                    if constexpr (!AllowsDuplicates) { running = false; }
+                    equals[i] = true;
+                    break;
+                  }
+                  default: {
+                    break;
+                  }
                 }
               }
             }
 
-            if (active_flushing_tile.any(match_found)) {
-              auto const matching_tile = cg::binary_partition(active_flushing_tile, match_found);
-              // stage matches in shmem buffer
-              if (match_found) {
-                probe_buffers[flushing_tile_id][num_matches + matching_tile.thread_rank()] = probe;
-                match_buffers[flushing_tile_id][num_matches + matching_tile.thread_rank()] =
-                  bucket_slots[i];
+            probing_tile.sync();
+            running = probing_tile.all(running);
+#pragma unroll buffer_size
+            for (int32_t i = 0; i < bucket_size; ++i) {
+              exists[i] = probing_tile.ballot(equals[i]);
+            }
+
+            // Fill the buffer if any matching keys are found
+            auto const lane_id = probing_tile.thread_rank();
+            if (thrust::any_of(thrust::seq, exists, exists + bucket_size, thrust::identity{})) {
+              if constexpr (IsOuter) { found_match = true; }
+
+              int32_t num_matches[bucket_size];
+
+              for (int32_t i = 0; i < bucket_size; ++i) {
+                num_matches[i] = __popc(exists[i]);
               }
 
-              // add number of new matches to the buffer counter
-              num_matches += (match_found) ? matching_tile.size()
-                                           : active_flushing_tile.size() - matching_tile.size();
-            }
+              auto const total_matches =
+                thrust::reduce(thrust::seq, num_matches, num_matches + bucket_size);
```

Comment:
Curious on your choice of thrust primitives here. Is this basically an `atomicAdd`?

Reply:
This is a sum reduction over the `num_matches` array. Each thread calculates the same sum.

Reply:
This and the below atomic basically do the same as in https://github.com/NVIDIA/cuCollections/blob/93d61722e7c04f6504a7707830864affc03e1d0e/include/cuco/detail/static_multimap/device_view_impl.inl#L1388-L1395

Now that you mention it, `total_matches` is only needed by thread 0. Let me clean this up.

Reply:
Fixed in 703da6a11ba8d01cda3460b4453926348a407548